src/Symfony/Component/Messenger/Transport/Serialization/Serializer.php

```diff
@@ -42,17 +44,37 @@ public function decode(array $encodedMessage)
             throw new \InvalidArgumentException('Encoded message does not have a `type` header.');
         }
 
-        return $this->serializer->deserialize($encodedMessage['body'], $encodedMessage['headers']['type'], $this->format, $this->context);
+        $configurations = isset($encodedMessage['configurations']) ? unserialize($encodedMessage['configurations']) : array();
+
+        $context = $this->context;
+        /** @var SerializerConfiguration|null $serializerConfig */
+        if ($serializerConfig = $configurations[SerializerConfiguration::class] ?? null) {
+            $context = $serializerConfig->getContext() + $context;
+        }
+
+        $message = $this->serializer->deserialize($encodedMessage['body'], $encodedMessage['headers']['type'], $this->format, $context);
+
+        return $configurations ? new ConfiguredMessage($message, $configurations) : $message;
     }
 
     /**
      * {@inheritdoc}
      */
     public function encode($message): array
     {
+        $configuredMessage = ConfiguredMessage::from($message);
+        $originalMessage = $configuredMessage->getMessage();
+
+        $context = $this->context;
+        /** @var SerializerConfiguration|null $serializerConfig */
+        if ($serializerConfig = $configuredMessage->get(SerializerConfiguration::class)) {
+            $context = $serializerConfig->getContext() + $context;
+        }
+
         return array(
-            'body' => $this->serializer->serialize($message, $this->format, $this->context),
-            'headers' => array('type' => \get_class($message)),
+            'body' => $this->serializer->serialize($originalMessage, $this->format, $context),
+            'headers' => array('type' => \get_class($originalMessage)),
+            'configurations' => $message instanceof ConfiguredMessage ? serialize($message->all()) : null,
```

Comment:
This won't necessarily be picked up by the adapters. I suggest we use headers for that. Typically, we have have the headers looking like that:
```
X-Message-Configuration-1: Symfony\Component\Messenger\Transport\Serialization\SerializerConfiguration:{"context": {"foo": "bar"}}
X-Message-Configuration-2: Symfony\Component\Messenger\Asynchronous\Transport\ReceivedMessage:{}
```

And we use Symfony Serializer to use some JSON so we ensure it's compatible with other systems reading it.